### PR TITLE
Default args type check

### LINKED ISSUE #21768: raise TypeError when default_args not a dictionary
### Apache Airflow version

2.2.4 (latest released)

### What happened

When triggering this dag below it runs when it should fail. A set is being passed to default_args instead of what should be a dictionary yet the dag still succeeds.

### What you expected to happen

I expected the dag to fail as the default_args parameter should only be a dictionary.

### How to reproduce

```
from airflow.models import DAG
from airflow.operators.python import PythonVirtualenvOperator, PythonOperator
from airflow.utils.dates import days_ago


def callable1():
   pass

with DAG(
    dag_id="virtualenv_python_operator",
    default_args={"owner: airflow"},
    schedule_interval=None,
    start_date=days_ago(2),
    tags=["core"],
) as dag:

    task = PythonOperator(
        task_id="check_errors",
        python_callable=callable1,
    )

```

### Operating System

Docker (debian:buster)

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

Astro CLI with images:
- quay.io/astronomer/ap-airflow-dev:2.2.4-1-onbuild
- quay.io/astronomer/ap-airflow-dev:2.2.3-2
- quay.io/astronomer/ap-airflow-dev:2.2.0-5-buster-onbuild

### Anything else

Bug happens every time.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by rustikk**
closes: https://github.com/apache/airflow/issues/21768

When the default_args is anything but a dictionary or None a not so helpful error shows. This commit adds a TypeError if anything other than a dictionary or None is passed to default_args at either the dag level, the taskgroup level or the task level and returns a TypeError saying "default_args must be a dictionary".

### CODE REVIEW

File: airflow/models/baseoperator.py
Comment (uranusjr) [inline airflow/models/baseoperator.py]:
```suggestion
        if task_group.default_args and not isinstance(task_group.default_args, collections.abc.Mapping):
            raise TypeError("default_args must be a mapping")
        dag_args.update(task_group.default_args)
```

Technically `default_args` does not need to be a dict, but only needs a [_dict-like interface_](https://docs.python.org/3/glossary.html#term-mapping).

File: airflow/models/baseoperator.py
Comment (uranusjr) [inline airflow/models/baseoperator.py]: Same for this.

File: airflow/models/dag.py
Comment (uranusjr) [inline airflow/models/dag.py]: This is the only place `default_args` must be a dict, but please use the built-in class instead of the typing alias.

```suggestion
        if default_args and not isinstance(default_args, dict):
            raise TypeError("default_args must be a dict")
        self.default_args = copy.deepcopy(default_args or {})
```

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/models/baseoperator.py
Comment (rustikk) [reply airflow/models/baseoperator.py]: @uranusjr good to know thank you for reviewing this.

**Comment by rustikk** [comment]
hey @uranusjr I've made all the changes you suggested and all of the checks have passed.

**Review by uranusjr — APPROVED** [review]

**Comment by rustikk** [comment]
@uranusjr The checks seems to be failing for an unrelated issue. What do I need to do next to get this merged into the main branch? Sorry I'm kind of new to working on open source projects.

**Comment by uranusjr** [comment]
The failures are fine, they are unrelated. Now it’s mostly patience waiting for another maintainer to chim in and review.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
No failures any more. And Approval there.

**Pull request merged**
